File: outputs/audits/auto_reference_20260711/qa-system/main.js
#!/usr/bin/env node
/**
 * 自动质检系统 - 主入口
 *
 * 用法:
 *   # 1. 初始化配置
 *   node main.js --setup
 *
 *   # 2. 单条质检
 *   node main.js --title "你的题目文本" --task-type L2
 *
 *   # 3. 批量质检（从 Excel 读取待检题目）
 *   node main.js --batch
 *
 *   # 4. 批量质检（限定数量）
 *   node main.js --batch --limit 10
 *
 *   # 5. 交互模式
 *   node main.js --interactive
 */
const fs = require('fs');
const path = require('path');
const readline = require('readline');
const config = require('./config');
const { QAEngine } = require('./qa_engine');
const { loadTestSet } = require('./data_loader');

// ========== 辅助函数 ==========

function checkApiKey() {
  const key = config.DEEPSEEK_API_KEY;
  if (!key || key === 'sk-your-api-key' || key === 'sk-your-deepseek-api-key-here') {
    console.log('='.repeat(60));
    console.log('❌ 请先配置 DeepSeek API Key！');
    console.log('='.repeat(60));
    console.log('');
    console.log('两种方式任选一种：');
    console.log('  1. 运行: node main.js --setup');
    console.log('  2. 编辑 qa-system/.env 文件，设置: DEEPSEEK_API_KEY=sk-xxx');
    console.log('');
    console.log('获取 API Key: https://platform.deepseek.com/api_keys');
    console.log('='.repeat(60));
    return false;
  }
  return true;
}

function parseArgs() {
  const args = {
    title: '',
    taskType: '',
    attachment: '',
    output: '',
    batch: false,
    limit: 0,
    interactive: false,
    noFewshot: false,
    quiet: false,
    setup: false,
    model: '',
  };

  const argv = process.argv.slice(2);
  for (let i = 0; i < argv.length; i++) {
    switch (argv[i]) {
      case '--title': args.title = argv[++i] || ''; break;
      case '--task-type': args.taskType = argv[++i] || ''; break;
      case '--attachment': args.attachment = argv[++i] || ''; break;
      case '--output': args.output = argv[++i] || ''; break;
      case '--batch': args.batch = true; break;
      case '--limit': args.limit = parseInt(argv[++i]) || 0; break;
      case '--interactive': case '-i': args.interactive = true; break;
      case '--no-fewshot': args.noFewshot = true; break;
      case '--quiet': case '-q': args.quiet = true; break;
      case '--setup': args.setup = true; break;
      case '--model': args.model = argv[++i] || ''; break;
      case '--help': case '-h':
        printHelp();
        process.exit(0);
    }
  }
  return args;
}

function printHelp() {
  console.log(`
🤖 自动质检系统 - 基于 DeepSeek V4 API

用法:
  node main.js [选项]

单条质检:
  node main.js --title "题目文本" --task-type L2

批量质检:
  node main.js --batch                  # 质检所有待检题目
  node main.js --batch --limit 10       # 只质检 10 条

交互模式:
  node main.js --interactive

初始化:
  node main.js --setup                  # 配置 API Key

选项:
  --title <text>       单条质检的题目文本
  --task-type <type>   任务类型 (L1/L2/L3)
  --attachment <text>  附件描述
  --output <text>      产物内容
  --batch              批量质检（从反馈表读取）
  --limit <n>          批量质检数量限制
  --interactive, -i    交互模式
  --no-fewshot         禁用 Few-Shot 案例检索（只靠规则判断）
  --quiet, -q          安静模式
  --model <name>       模型选择 (deepseek-chat / deepseek-reasoner)
  --setup              初始化配置
  --help, -h           显示帮助
`);
}

// ========== 命令实现 ==========

async function cmdSetup() {
  const envPath = path.join(__dirname, '.env');

  if (fs.existsSync(envPath)) {
    console.log(`⚠️ .env 已存在: ${envPath}`);
    const answer = await ask(`是否覆盖? (y/n): `);
    if (answer.toLowerCase() !== 'y') return;
  }

  const apiKey = await ask('请输入 DeepSeek API Key: ');
  const model = await ask('模型选择 (默认 deepseek-chat, 推理版用 deepseek-reasoner): ');
  const finalModel = model.trim() || 'deepseek-chat';

  const content = `# DeepSeek API 配置
DEEPSEEK_API_KEY=${apiKey.trim()}
DEEPSEEK_BASE_URL=https://api.deepseek.com
DEEPSEEK_MODEL=${finalModel}

# 文件路径（一般不需要改）
RULES_PATH=../垂域高难度题目生产--二期要求.md
PROMPT_PATH=../完整AI质检Prompt_QA_PROMPT原文_1.txt
CASES_PATH=../反馈表.xlsx
`;

  fs.writeFileSync(envPath, content, 'utf-8');
  console.log(`✅ 配置文件已创建: ${envPath}`);
}

function ask(question) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise(resolve => {
    rl.question(question, answer => { rl.close(); resolve(answer); });
  });
}

async function cmdSingle(args) {
  if (!checkApiKey()) return;

  const engine = new QAEngine();
  if (args.model) engine.model = args.model;

  const result = await engine.checkOne(
    args.title,
    args.taskType,
    args.attachment,
    args.output,
    !args.noFewshot
  );

  console.log('');
  console.log('='.repeat(60));
  console.log('📋 质检结果');
  console.log('='.repeat(60));

  if (result.error) {
    console.log(`❌ API 调用失败: ${result.error}`);
    return;
  }

  if (result.pass === true) {
    console.log('✅ 通过 —— 该题目无问题');
  } else if (result.pass === false) {
    console.log(`❌ 不通过 —— 发现 ${result.issues.length} 个问题：\n`);
    result.issues.forEach((issue, i) => {
      console.log(`  [${i + 1}] ${issue.category || ''} (${issue.severity || ''})`);
      console.log(`      问题: ${issue.problem || ''}`);
      console.log(`      建议: ${issue.short_note || issue.fix || ''}`);
      console.log('');
    });
  } else {
    console.log('⚠️ 结果解析失败，原始返回：');
    console.log((result.rawResponse || '').substring(0, 500));
  }

  console.log(`⏱️ 耗时: ${result.time}s | 💰 Token: ${result.tokens}`);
}

async function cmdBatch(args) {
  if (!checkApiKey()) return;

  console.log('📂 从反馈表加载待检题目...');
  const items = loadTestSet(config.CASES_PATH, args.limit);

  if (items.length === 0) {
    console.log('⚠️ 没有找到待质检的题目（所有题目都已有质检意见）');
    return;
  }

  console.log(`📋 共 ${items.length} 条待检题目\n`);

  const engine = new QAEngine();
  if (args.model) engine.model = args.model;

  const results = await engine.checkBatch(items, !args.noFewshot, !args.quiet);

  // 保存结果
  const jsonPath = path.join(config.OUTPUT_DIR, 'batch_result.json');
  engine.saveResults(results, jsonPath);

  // 生成报告
  generateReport(results, path.join(config.OUTPUT_DIR, 'batch_report.md'));
}

async function cmdInteractive(args) {
  if (!checkApiKey()) return;

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

  console.log('');
  console.log('='.repeat(60));
  console.log('🤖 自动质检系统 - 交互模式');
  console.log('   输入题目文本后按 Enter 提交质检');
  console.log('   输入 quit 退出');
  console.log('='.repeat(60));

  const engine = new QAEngine();
  if (args.model) engine.model = args.model;

  const askQ = (q) => new Promise(resolve => rl.question(q, resolve));

  while (true) {
    console.log('');
    console.log('-'.repeat(40));
    const taskType = await askQ('任务类型 (L1/L2/L3，可选): ');
    if (taskType.toLowerCase() === 'quit') break;

    const title = await askQ('题目文本: ');
    if (title.toLowerCase() === 'quit') break;

    if (!title.trim()) {
      console.log('题目为空，跳过');
      continue;
    }

    console.log('\n🔍 正在质检...');
    const result = await engine.checkOne(title, taskType);

    if (result.pass === true) {
      console.log('✅ 通过');
    } else if (result.pass === false) {
      console.log(`❌ 不通过 - ${result.issues.length} 个问题`);
      result.issues.forEach(issue => {
        console.log(`  - [${issue.category}] ${issue.problem}`);
        console.log(`    > ${issue.short_note || issue.fix || ''}`);
      });
    } else {
      console.log(`⚠️ 异常: ${result.error || (result.rawResponse || '').substring(0, 200)}`);
    }
  }

  rl.close();
}

function generateReport(results, outputPath) {
  const lines = [];
  const now = new Date().toISOString().replace('T', ' ').substring(0, 19);

  lines.push('# 自动质检报告');
  lines.push('');
  lines.push(`**质检时间**: ${now}`);
  lines.push(`**总数**: ${results.length}`);
  lines.push('');

  const passed = results.filter(r => r.result.pass === true).length;
  const failed = results.filter(r => r.result.pass === false).length;
  const errorCount = results.filter(r => r.result.pass === null).length;

  lines.push('| 状态 | 数量 |');
  lines.push('|------|------|');
  lines.push(`| ✅ 通过 | ${passed} |`);
  lines.push(`| ❌ 不通过 | ${failed} |`);
  lines.push(`| ⚠️ 异常 | ${errorCount} |`);
  lines.push('');
  lines.push('---');
  lines.push('');

  for (const item of results) {
    const result = item.result;
    const status = result.pass === true ? '✅' : result.pass === false ? '❌' : '⚠️';
    lines.push(`## ${status} ${item.uid || ''}`);
    lines.push(`- **题目**: ${(item.title || '').substring(0, 200)}`);
    lines.push(`- **任务类型**: ${item.taskType || ''}`);

    if (result.issues && result.issues.length > 0) {
      lines.push('- **问题**:');
      for (const issue of result.issues) {
        lines.push(`  - ${issue.category || ''}: ${issue.problem || ''}`);
        lines.push(`    > ${issue.short_note || issue.fix || ''}`);
      }
    }

    if (result.error) {
      lines.push(`- **错误**: ${result.error}`);
    }

    lines.push(`- Token: ${result.tokens || 0} | 耗时: ${result.time || 0}s`);
    lines.push('');
  }

  const dir = path.dirname(outputPath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  fs.writeFileSync(outputPath, lines.join('\n'), 'utf-8');
  console.log(`📄 报告已保存到: ${path.resolve(outputPath)}`);
}

// ========== 主函数 ==========

async function main() {
  const args = parseArgs();

  if (args.setup) {
    await cmdSetup();
    return;
  }

  if (args.batch) {
    await cmdBatch(args);
  } else if (args.interactive) {
    await cmdInteractive(args);
  } else if (args.title) {
    await cmdSingle(args);
  } else {
    printHelp();
    console.log('💡 提示: 使用 --batch 进行批量质检，或 --title 进行单条质检');
  }
}

main().catch(err => {
  console.error('❌ 运行出错:', err.message);
  process.exit(1);
});

File: outputs/audits/auto_reference_20260711/qa-system/main.py
#!/usr/bin/env python3
"""
自动质检系统 - 主入口
用法:
    # 1. 设置 API Key（二选一）
    #    方式A: 复制 .env.example 为 .env，填入你的 DeepSeek API Key
    #    方式B: 设置环境变量 export DEEPSEEK_API_KEY=sk-xxx

    # 2. 单条质检
    python main.py --title "你的题目文本" --task-type L2

    # 3. 批量质检（从 Excel 读取待检题目）
    python main.py --batch

    # 4. 批量质检（限定数量）
    python main.py --batch --limit 10

    # 5. 从文件读取题目质检
    python main.py --file test_questions.txt

    # 6. 交互模式
    python main.py --interactive
"""
import sys
import os
import argparse
import json
from pathlib import Path

# 确保能导入同目录模块
sys.path.insert(0, str(Path(__file__).parent))

from config import (
    DEEPSEEK_API_KEY, RULES_PATH, PROMPT_PATH, CASES_PATH, OUTPUT_DIR
)
from qa_engine import QAEngine


def check_api_key():
    """检查 API Key 是否已配置"""
    if DEEPSEEK_API_KEY in ("sk-your-api-key", "sk-your-deepseek-api-key-here", ""):
        print("=" * 60)
        print("❌ 请先配置 DeepSeek API Key！")
        print("=" * 60)
        print("\n两种方式任选一种：")
        print("  1. 编辑 qa-system/.env 文件，设置: DEEPSEEK_API_KEY=sk-xxx")
        print("  2. 设置环境变量: set DEEPSEEK_API_KEY=sk-xxx")
        print("\n获取 API Key: https://platform.deepseek.com/api_keys")
        print("=" * 60)
        return False
    return True


def cmd_single(args):
    """单条质检"""
    if not check_api_key():
        return

    engine = QAEngine(prompt_path=PROMPT_PATH, cases_path=CASES_PATH)

    result = engine.check_one(
        title=args.title,
        task_type=args.task_type or "",
        attachment_details=args.attachment or "",
        output=args.output or "",
        use_few_shot=not args.no_fewshot,
    )

    print("\n" + "=" * 60)
    print("📋 质检结果")
    print("=" * 60)

    if result.get("error"):
        print(f"❌ API 调用失败: {result['error']}")
        return

    if result["pass"]:
        print("✅ 通过 —— 该题目无问题")
    elif result["pass"] is False:
        print(f"❌ 不通过 —— 发现 {len(result['issues'])} 个问题：\n")
        for i, issue in enumerate(result["issues"], 1):
            print(f"  [{i}] {issue.get('category', '')} ({issue.get('severity', '')})")
            print(f"      问题: {issue.get('problem', '')}")
            print(f"      建议: {issue.get('short_note', issue.get('fix', ''))}")
            print()
    else:
        print("⚠️ 结果解析失败，原始返回：")
        print(result.get("raw_response", "")[:500])

    print(f"⏱️ 耗时: {result['time']}s | 💰 Token: {result['tokens']}")


def cmd_batch(args):
    """批量质检"""
    if not check_api_key():
        return

    from data_loader import load_test_set

    print(f"📂 从反馈表加载待检题目...")
    items = load_test_set(CASES_PATH, limit=args.limit)

    if not items:
        print("⚠️ 没有找到待质检的题目（所有题目都已有质检意见）")
        return

    print(f"📋 共 {len(items)} 条待检题目\n")

    engine = QAEngine(prompt_path=PROMPT_PATH, cases_path=CASES_PATH)

    results = engine.check_batch(
        items,
        use_few_shot=not args.no_fewshot,
        verbose=not args.quiet,
    )

    # 保存结果
    output_path = OUTPUT_DIR / "batch_result.json"
    engine.save_results(results, str(output_path))

    # 同时生成可读报告
    from data_loader import load_cases_from_excel
    _generate_report(results, OUTPUT_DIR / "batch_report.md")


def cmd_interactive(args):
    """交互模式"""
    if not check_api_key():
        return

    print("\n" + "=" * 60)
    print("🤖 自动质检系统 - 交互模式")
    print("   输入题目文本，按 Enter 后 Ctrl+D (Unix) 或 Ctrl+Z (Win) 提交")
    print("   输入 'quit' 退出")
    print("=" * 60)

    engine = QAEngine(prompt_path=PROMPT_PATH, cases_path=CASES_PATH)

    while True:
        print("\n" + "-" * 40)
        task_type = input("任务类型 (L1/L2/L3，可选): ").strip()

        if task_type.lower() == "quit":
            break

        print("题目文本（支持多行，输入 'done' 结束）:")
        lines = []
        while True:
            try:
                line = input()
            except EOFError:
                break
            if line.strip() == "done":
                break
            lines.append(line)

        title = "\n".join(lines)
        if not title.strip():
            print("题目为空，跳过")
            continue

        print("\n🔍 正在质检...")
        result = engine.check_one(title=title, task_type=task_type)

        if result.get("pass"):
            print("✅ 通过")
        elif result.get("pass") is False:
            print(f"❌ 不通过 - {len(result['issues'])} 个问题")
            for issue in result["issues"]:
                print(f"  - {issue.get('problem', '')}")
        else:
            print(f"⚠️ 解析异常: {result.get('raw_response', '')[:300]}")


def _generate_report(results: list, output_path: str):
    """生成可读的 Markdown 报告"""
    lines = [
        "# 自动质检报告",
        "",
        f"**质检时间**: {__import__('datetime').datetime.now().strftime('%Y-%m-%d %H:%M:%S')}",
        f"**总数**: {len(results)}",
        "",
        f"| 状态 | 数量 |",
        f"|------|------|",
    ]

    passed = sum(1 for r in results if r["result"].get("pass"))
    failed = sum(1 for r in results if r["result"].get("pass") is False)
    error_count = sum(1 for r in results if r["result"].get("pass") is None)

    lines.append(f"| ✅ 通过 | {passed} |")
    lines.append(f"| ❌ 不通过 | {failed} |")
    lines.append(f"| ⚠️ 异常 | {error_count} |")
    lines.append("")
    lines.append("---")
    lines.append("")

    for item in results:
        result = item["result"]
        status = "✅" if result.get("pass") else ("❌" if result.get("pass") is False else "⚠️")
        lines.append(f"## {status} {item.get('uid', '')}")
        lines.append(f"- **题目**: {item.get('title', '')[:200]}")
        lines.append(f"- **任务类型**: {item.get('task_type', '')}")

        if result.get("issues"):
            lines.append("- **问题**:")
            for issue in result["issues"]:
                lines.append(f"  - {issue.get('category', '')}: {issue.get('problem', '')}")
                lines.append(f"    > {issue.get('short_note', issue.get('fix', ''))}")

        if result.get("error"):
            lines.append(f"- **错误**: {result['error']}")

        lines.append(f"- Token: {result.get('tokens', 0)} | 耗时: {result.get('time', 0)}s")
        lines.append("")

    with open(output_path, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))
    print(f"📄 报告已保存到: {output_path}")


def main():
    parser = argparse.ArgumentParser(
        description="🤖 自动质检系统 - 基于 DeepSeek V4 API",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:
  python main.py --title "帮我制定一份旅游计划" --task-type L1
  python main.py --batch --limit 5
  python main.py --interactive
        """
    )

    # 单条质检
    parser.add_argument("--title", type=str, help="单条质检：题目文本")
    parser.add_argument("--task-type", type=str, default="", help="任务类型 (L1/L2/L3)")
    parser.add_argument("--attachment", type=str, default="", help="附件描述")
    parser.add_argument("--output", type=str, default="", help="产物内容")

    # 批量质检
    parser.add_argument("--batch", action="store_true", help="批量质检：从反馈表读取待检题目")
    parser.add_argument("--limit", type=int, default=0, help="批量质检数量限制（0=全部）")

    # 交互模式
    parser.add_argument("--interactive", "-i", action="store_true", help="交互模式")

    # 选项
    parser.add_argument("--no-fewshot", action="store_true", help="禁用 Few-Shot 案例检索")
    parser.add_argument("--quiet", "-q", action="store_true", help="安静模式（批量时不打印进度）")
    parser.add_argument("--model", type=str, default="", help="覆盖模型选择")

    # 配置
    parser.add_argument("--setup", action="store_true", help="初始化 .env 配置文件")

    args = parser.parse_args()

    if args.setup:
        _setup_env()
        return

    # 检查模型覆盖
    if args.model:
        import config
        config.DEEPSEEK_MODEL = args.model

    if args.batch:
        cmd_batch(args)
    elif args.interactive:
        cmd_interactive(args)
    elif args.title:
        cmd_single(args)
    else:
        parser.print_help()
        print("\n💡 提示: 使用 --batch 进行批量质检，或 --title 进行单条质检")


def _setup_env():
    """创建 .env 配置文件"""
    env_path = Path(__file__).parent / ".env"
    example_path = Path(__file__).parent / ".env.example"

    if env_path.exists():
        print(f"⚠️ .env 已存在: {env_path}")
        overwrite = input("是否覆盖? (y/n): ").strip().lower()
        if overwrite != "y":
            return

    api_key = input("请输入 DeepSeek API Key: ").strip()
    model = input("模型选择 (默认 deepseek-chat, 推理版用 deepseek-reasoner): ").strip()
    if not model:
        model = "deepseek-chat"

    with open(env_path, "w", encoding="utf-8") as f:
        f.write(f"""# DeepSeek API 配置
DEEPSEEK_API_KEY={api_key}
DEEPSEEK_BASE_URL=https://api.deepseek.com
DEEPSEEK_MODEL={model}

# 文件路径
RULES_PATH=../垂域高难度题目生产--二期要求.md
PROMPT_PATH=../完整AI质检Prompt_QA_PROMPT原文_1.txt
CASES_PATH=../反馈表.xlsx
""")
    print(f"✅ 配置文件已创建: {env_path}")


if __name__ == "__main__":
    main()
